scripts/db/src/login.ts

```diff
@@ -37,52 +59,124 @@ async function login(
+      const forumUser = await mongoClient
+        .db(MONGO_DB_NAME)
+        .collection<ForumUser>('users')
+        .findOne({ 'emails.address': email })
+
+      await mongoClient.close()
+
+      if (!forumUser) {
+        return null
+      }
+
+      // Check their password
+      // Meteor hashed its passwords twice, once on the client and once again on
+      // the server. To preserve backwards compatibility with Meteor passwords,
+      // we do the same, but do it both on the server-side
+      const meteorClientSideHash = createHash('sha256')
+        .update(password)
+        .digest('hex')
+      const isValid = await bcrypt.compare(
+        meteorClientSideHash,
+        forumUser.services.password.bcrypt
+      )
+      if (!isValid) {
+        return null
+      }
+
+      // Which email did we find?
+      const emailInfo = forumUser.emails.find((e) => e.address === email)
+      if (!emailInfo) {
+        // This should never happen, as they were returned by mongo because that
+        // field matched
+        throw new Error(
+          `User found by email ${email}, does not have that email (should never happen)`
+        )
+      }
+
+      return {
+        id: forumUser._id,
+        username: forumUser.displayName,
+        email: emailInfo.address,
+        email_verified: emailInfo.verified,
+      }
     }
 
-    /** Construct a postgres client and connect to the server */
-    const client: Client = new Client(connectionInfo)
-    await client.connect()
+    async function loginParfitUser(): Promise<CallbackUser | null> {
+      /** Declare connection info */
+      const pgConnectionInfo: PGConnectionConfig = {
+        user: POSTGRES_USERNAME,
+        password: POSTGRES_PASSWORD,
+        host: POSTGRES_HOST,
+        database: POSTGRES_DATABASE,
+        port: POSTGRES_PORT ? parseInt(POSTGRES_PORT) : 5432,
+        ssl: TEMPLATE_DATA.pgShouldSsl,
+      }
+      /** Construct a postgres client and connect to the server */
+      const pgClient: PGClient = new PGClient(pgConnectionInfo)
+      await pgClient.connect()
 
-    /** Get the person based on their email, joining on the password table */
-    const query = `
+      /** Get the person based on their email, joining on the password table */
+      const parfitQuery = `
       select
         person.id, email, first_name, last_name, password
       from people.person
       join auth.password on password.person_id = person.id
       where person.email = $1
     `
-    const result = await client.query<PersonResult>(query, [email])
+      const parfitResult = await pgClient.query<PersonResult>(parfitQuery, [
+        email,
+      ])
 
-    /** Close the connection */
-    await client.end()
+      /** Close the connection */
+      await pgClient.end()
 
-    if (result.rows.length === 0) {
-      return callback(new WrongUsernameOrPasswordError(email))
-    }
+      const Person = parfitResult.rows[0]
```

Comment:
This seems like it should come after the block below, where you inspect the `parfitResult`.

Also, a weak preference (no real reason other than habit, though it does have the nice feature of not requiring an intermediate variable) for dealing with single-result queries like this:

```ts
const Person: ParfitPerson = await pgClient.query<PersonResult>(parfitQuery, [
	email,
]).then(res => res.rows[0])

if (!Person) {
	return null
}
```

Reply:
(Also, should this have a type declaration, e.g. `const Person: ParfitPerson`)

Reply:
This is you commenting on your own previous Auth0 code, but I've done it.

And the type declaration is already present because of the specification in the generic function.